File: Framework/CSharp/Framework/Framework/SmartCompare.cs
/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */
using System;
using System.Collections.Generic;

namespace Smartkernel.Framework
{
	/// <summary>
	/// 比较
	/// </summary>
	public static class SmartCompare
	{
		/// <summary>
		/// 判断一个值是否在一个值范围内
		/// </summary>
		/// <param name="input">输入</param>
		/// <param name="start">开始</param>
		/// <param name="end">结束</param>
		/// <returns>结果</returns>
		public static bool InScope<T>(T input, T start, T end) where T :IComparable
		{
			return input.CompareTo(start) >= 0 && input.CompareTo(end) <= 0;
		}

		/// <summary>
		/// 判断一个值是否不在一个值范围内
		/// </summary>
		/// <param name="input">输入</param>
		/// <param name="start">开始</param>
		/// <param name="end">结束</param>
		/// <returns>结果</returns>
		public static bool NotInScope<T>(T input, T start, T end) where T : IComparable
		{
			return !InScope(input, start, end);
		}

		/// <summary>
		/// 判断一个值是否在一个值范围内（有一个在其范围内即为真）
		/// </summary>
		/// <param name="inputs">输入</param>
		/// <param name="start">开始</param>
		/// <param name="end">结束</param>
		/// <returns>结果</returns>
		public static bool AnyInScope<T>(List<T> inputs, T start, T end) where T : IComparable
		{
			foreach (var input in inputs)
			{
				if (InScope<T>(input,start,end))
				{
					return true;
				}
			}
			return false;
		}

		/// <summary>
		/// 判断一个值是否不在一个值范围内（有一个在其范围内即为假）
		/// </summary>
		/// <param name="inputs">输入</param>
		/// <param name="start">开始</param>
		/// <param name="end">结束</param>
		/// <returns>结果</returns>
		public static bool NotAnyInScope<T>(List<T> inputs, T start, T end) where T : IComparable
		{
			return !AnyInScope(inputs, start, end);
		}
		
		/// <summary>
		/// 判断一个值是否在一个值范围内（有一个在其范围内即为真）
		/// </summary>
		/// <param name="inputs">输入</param>
		/// <param name="start">开始</param>
		/// <param name="end">结束</param>
		/// <returns>结果</returns>
		public static bool AllInScope<T>(List<T> inputs, T start, T end) where T : IComparable
		{
			foreach (var input in inputs)
			{
				if (!InScope<T>(input,start,end))
				{
					return false;
				}
			}
			return true;
		}

		/// <summary>
		/// 判断一个值是否不在一个值范围内（有一个在其范围内即为假）
		/// </summary>
		/// <param name="inputs">输入</param>
		/// <param name="start">开始</param>
		/// <param name="end">结束</param>
		/// <returns>结果</returns>
		public static bool NotAllInScope<T>(List<T> inputs, T start, T end) where T : IComparable
		{
			return !AllInScope(inputs, start, end);
		}
	}
}

File: Php/framework/framework/smartcompare.php
<?php

namespace Smartkernel\Framework {
    /*
     * 比较
     */

    class SmartCompare {
        /*
         * 判断一个值是否在一个值范围内
         * 
         * @param number $input，输入
         * @param number $start，开始
         * @param number $end，结束
         * @return bool，结果
         * 
         */

        public static function inScope($input, $start, $end) {
            return $input >= $start && $input <= $end;
        }

        /*
         * 判断一个值是否不在一个值范围内
         * 
         * @param number $input，输入
         * @param number $start，开始
         * @param number $end，结束
         * @return bool，结果
         * 
         */

        public static function notInScope($input, $start, $end) {
            return !self::inScope($input, $start, $end);
        }

        /*
         * 判断一个值是否在一个值范围内（有一个在其范围内即为真）
         * 
         * @param array $inputs，输入
         * @param number $start，开始
         * @param number $end，结束
         * @return bool，结果
         * 
         */

        public static function anyInScope(array $inputs, $start, $end) {
            foreach ($inputs as $input) {
                if (self::inScope($input, $start, $end)) {
                    return true;
                }
            }
            return false;
        }

        /*
         * 判断一个值是否不在一个值范围内（有一个在其范围内即为假）
         * 
         * @param array $inputs，输入
         * @param number $start，开始
         * @param number $end，结束
         * @return bool，结果
         * 
         */

        public static function notAnyInScope(array $inputs, $start, $end) {
            return self::anyInScope($inputs, $start, $end);
        }

        /*
         * 判断一个值是否在一个值范围内（有一个在其范围内即为真）
         * 
         * @param array $inputs，输入
         * @param number $start，开始
         * @param number $end，结束
         * @return bool，结果
         * 
         */

        public static function allInScope(array $inputs, $start, $end) {
            foreach ($inputs as $input) {
                if (!self::inScope($input, $start, $end)) {
                    return false;
                }
            }
            return true;
        }

        /*
         * 判断一个值是否不在一个值范围内（有一个在其范围内即为假）
         * 
         * @param array $inputs，输入
         * @param number $start，开始
         * @param number $end，结束
         * @return bool，结果
         * 
         */

        public static function notAllInScope(array $inputs, $start, $end) {
            return self::allInScope($inputs, $start, $end);
        }

    }

}
